File: cpp/oop/inheritance.cpp
#include <iostream>
#include <string>

class BankAccount{
    protected:
        std::string account_num;
        double balance;
    
    public:
        BankAccount(std::string account_n, double bal) : account_num(account_n), balance(bal) {}
    
        std::string get_account_num(){
            return account_num;
        }

        double get_balance(){
            return balance;
        }
    
        void deposit(double amount){
            balance += amount;
            std::cout<<"Deposited "<<amount<<" to account "<<account_num<<". New balance: "<<balance<<std::endl;
        }
    
        void withdraw(double amount){
            if (amount <= balance) {
                balance -= amount;
                std::cout<<"Withdrew "<<amount<<" from account "<<account_num<<". New balance: "<<balance<<std::endl;
            } else {
                std::cout<<"Insufficient funds!"<<std::endl;
            }
        }
};

class SavingsAccount : public BankAccount{
    private : 
        float interest_rate;
    public : 
        SavingsAccount(std::string account_n, double balance, float int_rate=0.03) 
            : BankAccount(account_n, balance), interest_rate(int_rate) {}
    
        void add_interest(){
            double interest = balance * interest_rate;
            deposit(interest);
            std::cout<<"Interest of "<<interest<<" added to account "<<account_num<<". New balance: "<<balance<<std::endl;
        }
};

class CurrentAccount : public BankAccount{
    public:
        CurrentAccount(std::string account_n, double balance) 
            : BankAccount(account_n, balance) {}
};

int main(){
    SavingsAccount savings("SA1234567890", 1000.0);
    savings.deposit(500.0);
    savings.add_interest();
    savings.withdraw(300.0);
    std::cout << "Savings Account Balance: " << savings.get_balance() << std::endl<<std::endl;

    CurrentAccount current("CA1234567890", 200.0);
    current.deposit(300.0);
    current.withdraw(600.0);
    current.withdraw(200.0);  // This should exceed the overdraft limit
    std::cout << "Current Account Balance: " << current.get_balance() << std::endl;
    return 0;
};

File: python/oop/inheritance.py
class BankAccount:
    def __init__(self, account_num, balance):
        self._account_num = account_num
        self._balance = balance

    def get_account_num(self):
        return self._account_num

    def get_balance(self):
        return self._balance
    
    def deposit(self, amount):
        self._balance += amount
        print(f"Deposited {amount} to account {self._account_num}. New balance: {self._balance}")

    
    def withdraw(self, amount):
        if amount <= self._balance:
            self._balance -= amount
            print(f"Withdrew {amount} from account {self._account_num}. New balance: {self._balance}")
        else:
            print("Insufficient funds!")

class SavingsAccount(BankAccount):
    def __init__(self, account_num, balance, interest_rate=0.03):
        super().__init__(account_num, balance)
        self.interest_rate = interest_rate

    def add_interest(self):
        interest = self._balance * self.interest_rate
        self.deposit(interest)
        print(f"Interest of {interest} added to account {self._account_num}. New balance: {self._balance}")

class CurrentAccount(BankAccount):
    def __init__(self, account_num, balance):
        super().__init__(account_num, balance)

def main():
    savings = SavingsAccount("SA 1234 5678 90", 1000.0)
    savings.deposit(500.0)
    savings.add_interest()
    savings.withdraw(300.0)
    print(f"Savings Account Balance: {savings.get_balance()}\n")

    current = CurrentAccount("CA 1234 5678 90", 200.0)
    current.deposit(300.0)
    current.withdraw(600.0)
    current.withdraw(200.0) 
    print(f"Current Account Balance: {current.get_balance()}")


if __name__ == "__main__":
    main()